jinahub/encoders/text/CLIPTextEncoder/tests/unit/test_clip_text.py

```diff
@@ -16,7 +34,7 @@ def test_clip_batch():
 
 def test_clip_data():
```

Comment:
This test has way too many things (not your fault, of course). Let's refactor it.
- comparison with OpenAI's model should be a separate test (and make sure to implcitly set the model name for `CLIPTextEncoder` there)
- use the `DocumentArray.match` function instead of the custom `dist` function (see what I did [here](https://github.com/jina-ai/executors/pull/24/files#diff-5c95ffd73f009e4cf74526e5217d55c8a33b51def1d773619faa6fecd460e0f1R153-R176))
- Separate the comparison of sentences and words into two different tests
- For words comparison, replace the words list with `['apple', 'banana', 'satellite', 'airplane']` and check that fruits are similar to themselves, and airplane/satellite to each other (see again how i did it on link above)

Reply:
Hi,

What do you mean by implicitly set the model name? Mention model name explicitly as an argument to `CLIPTextEncoder` or use default value defined in the `__init__`?

I observed a funny thing w.r.t word embeddings:
 - Apple and banana are not treated as similar. I tested this with openai implementation and found the same behavior.
 - When I try with [apple1, apple2, banana1, banana2] then they are matched correctly.

Following is the code snippet:
```
def test_words_embedding():
    """
    Check if the distance between embeddings of similar words are smaller
    than dissimilar pair of words.
    """

    docs = DocumentArray(
        [
            Document(id="A", text="apple"),
            Document(id="B", text="banana"),
            Document(id="C", text="satellite"),
            Document(id="D", text="airplane"),
        ]
    )

    clip_text_encoder = CLIPTextEncoder()
    clip_text_encoder.encode(DocumentArray(docs), {})

    docs.match(docs, metric="euclidean")

    matches = ["B", "A", "D", "C"]
    for i, doc in enumerate(docs):
        assert doc.matches[1].id == matches[i]
```

Reply:
Hi @tadejsv,

Any suggestion on how to proceed with this?

Reply:
Hey, sorry I missed your reply yesterday. I'll look into this right away

Reply:
@makram93 I think I have solved this now. Probably the example strings are not the best - apple can be a fruit or a technology company, which then explains why it can be more similar to a satellite than to a banana. Let's replace the examples with `['dog', 'cat', 'airplane', 'helicopter']`.

Using just transformers this yields proper results for me
``` python
from transformers import CLIPProcessor, CLIPModel
from scipy.spatial.distance import pdist, squareform
import torch

model = CLIPModel.from_pretrained("openai/clip-vit-base-patch32")
processor = CLIPProcessor.from_pretrained("openai/clip-vit-base-patch32")

with torch.no_grad():
    inputs = processor(text=["cat", "dog", "helicopter", "airplane"], return_tensors="pt", padding=True)
    outputs = model.get_text_features(**inputs).numpy()

print(squareform(pdist(outputs, metric='cosine')))
```

```
[[0.         0.08779333 0.22051363 0.25879244]
 [0.08779333 0.         0.20850509 0.2353667 ]
 [0.22051363 0.20850509 0.         0.16761544]
 [0.25879244 0.2353667  0.16761544 0.        ]]
```

Reply:
Actually, let's change this a bit. Let the text be this list
```python
texts=[
    "a furry animal that with a long tail",
    "a domesticated mammal with four legs",
    "a type of aircraft that uses rotating, or spinning, wings",
    "flying vehicle that has fixed wings and engines",
]
```
I've noticed that performance on words is not as good as on sentences - which makes sense, because the model was trained on sentences.

Reply:
Ah yes, that makes sense why it's getting confused with the word `apple`. When I try with some other fruits' names, it works fine. Thanks for pointing this out. :)

I'll replace word embeddings with sentence embeddings and push the changes.

Reply:
Sorry, and to answer your other question: I meant setting the `CLIPTextEncoder` model name at initialization in that test, yes